### PR TITLE
hold: display edit buttons on all human messages

### PR DESCRIPTION
**Pull request by abeatrix**
close https://github.com/sourcegraph/cody/issues/2619

## This PR is outdated - moving to a new branch

The goal of this PR is to address the following items described in the design figma:

image

image

This PR includes the following changes:
- display edit button on each human message
- update the `handleEdit` function in SimpleChat to support commands and regular chat input properly
  - instead of only able to edit the last human message, user can now choose any human messages in their chat session to edit
  - also updated to support `/commands` in edited messages 

New tests added to cover the new changes
- update tests for the edit function in agent 
- add new e2e test for the updated user journey for editing chat messages

## Follow-up

- the new design update for the edit text area https://github.com/sourcegraph/cody/issues/2762

image

## Test plan

1. Build from this branch
2. in debug mode, ask Cody multiple questions, including using chat commands (e.g, /explain)
  - confirm edit button shows up on each message submitted when no message is in progress
3. click on an edit button to start the editing process
  - confirm the edit text area only shows up for the message you clicked the edit button on
4. update the message to something else and then submit the edit
  - confirm all the messages after the edited messages are removed (similar to how editing old messages work in chatGPT)
  - confirm the previous message has been replaced by the updated message
  - confirm cody is responding to your updated message when replying

Added new tests to agent to cover this updated behavior for edit. All tests are passing locally:

image

## Demo

https://github.com/sourcegraph/cody/assets/68532117/4a1b1cc5-f87f-4e2e-a491-412de82120b4

### CODE REVIEW

File: lib/ui/src/chat/Transcript.tsx [hunk 1]
```diff
@@ -156,12 +156,16 @@ export const Transcript: React.FunctionComponent<
             return (
                 <TranscriptItem
                     key={index + offset}
+                    index={index + offset - 1} // remove the welcome message
                     message={message}
                     inProgress={
                         offsetIndex && messageInProgress?.speaker === 'assistant' && !messageInProgress?.displayText
                     }
-                    beingEdited={messageBeingEdited && offsetIndex}
+                    showEditButton={!messageInProgress?.speaker && message.speaker === 'human'}
```
Comment (abeatrix) [inline lib/ui/src/chat/Transcript.tsx:164]: show edit buttons on all human messages

File: lib/ui/src/chat/TranscriptItem.tsx [hunk 2]
```diff
@@ -95,46 +97,62 @@ export const TranscriptItem: React.FunctionComponent<
...
+    // To identify if the current message is the one being edited when beingEdited is true
+    // This is used to display EditTextArea only for the message being edited
+    const [isItemBeingEdited, setIsItemBeingEdited] = useState<boolean>(false)
+
+    const onEditCurrentMessage = useCallback(
+        (status: boolean) => {
+            setBeingEdited(status)
+            setIsItemBeingEdited(status)
+        },
+        [setBeingEdited]
```
Comment (dominiccooney) [inline lib/ui/src/chat/TranscriptItem.tsx:115]: We should add tests where you start editing more than one message at a time, etc. if that is possible. (Would be nice if you had to cancel one edit before starting another...)

File: vscode/src/chat/chat-view/SimpleChatModel.ts [hunk 3]
```diff
@@ -107,6 +107,30 @@ export class SimpleChatModel {
...
+    public getLastSpeakerMessageIndex(speaker: 'human' | 'assistant'): number | undefined {
+        return this.messagesWithContext.findLastIndex(message => message.message.speaker === speaker)
+    }
+
+    /**
+     * Removes all messages from the given index only if it matches the expected speaker.
+     * expectedSpeaker is the expected speaker of the message at the given index
+     * to ensure the correct messages are being removed.
+     */
+    public removeMessagesFromIndex(index: number, expectedSpeaker: 'human' | 'assistant'): void {
```
Comment (dominiccooney) [inline vscode/src/chat/chat-view/SimpleChatModel.ts:119]: Why do we need `expectedSpeaker`? Seems like you're taking a later error, where the speakers have become inconsistent, and are bringing it closer to the likely bug. That's a good idea, but I'm curious why we can't go further.

File: vscode/src/chat/chat-view/SimpleChatModel.ts [hunk 4]
```diff
@@ -107,6 +107,30 @@ export class SimpleChatModel {
...
+        // nothing to remove
+        if (this.isEmpty()) {
+            return
+        }
+
+        const messageAtIndex = this.messagesWithContext.at(index)
+        if (messageAtIndex?.message?.speaker !== expectedSpeaker) {
+            throw new Error(`SimpleChatModel.removeHumanMessageByIndex: ${expectedSpeaker} is not the expected speaker`)
```
Comment (dominiccooney) [inline vscode/src/chat/chat-view/SimpleChatModel.ts:127]: This message is confusing, `${expectedSpeaker}` is the expected speaker. Just echo both, label which is which, and highlight that they are different.

File: vscode/src/chat/chat-view/SimpleChatPanelProvider.ts [hunk 5]
```diff
@@ -701,10 +701,25 @@ export class SimpleChatPanelProvider implements vscode.Disposable, ChatSession {
         this.updateWebviewPanelTitle(inputText)
     }
 
-    private async handleEdit(requestID: string, text: string): Promise<void> {
-        this.chatModel.updateLastHumanMessage({ text })
-        this.postViewTranscript()
-        await this.generateAssistantResponse(requestID)
+    /**
+     * TODO (bee) support userContextFiles and addEnhancedContext
```
Comment (dominiccooney) [inline vscode/src/chat/chat-view/SimpleChatPanelProvider.ts:705]: Let's do it?

File: vscode/src/chat/chat-view/SimpleChatPanelProvider.ts [hunk 6]
```diff
@@ -701,10 +701,25 @@ export class SimpleChatPanelProvider implements vscode.Disposable, ChatSession {
...
+     * Handles editing a previous chat message.
+     *
+     * Removes any existing messages at the provided index before submitting
```
Comment (dominiccooney) [inline vscode/src/chat/chat-view/SimpleChatPanelProvider.ts:708]: Would it be simpler to always take an index?

**Review by dominiccooney — CHANGES_REQUESTED** [review]
Nice start, let's try to simplify this a bit.

File: vscode/src/chat/chat-view/SimpleChatPanelProvider.ts (hunk 5 above)
Comment (abeatrix) [reply vscode/src/chat/chat-view/SimpleChatPanelProvider.ts:705]: for addEnhancedContext, it is done in https://github.com/sourcegraph/cody/pull/2806

File: lib/ui/src/chat/TranscriptItem.tsx [hunk 7]
```diff
@@ -95,46 +97,62 @@ export const TranscriptItem: React.FunctionComponent<
...
-                        setBeingEdited(false)
-                        editButtonOnSubmit(formInput)
-                    }}
-                    disabled={formInput.length === 0}
-                />
-            </div>
-        ) : null
+    // To identify if the current message is the one being edited when beingEdited is true
+    // This is used to display EditTextArea only for the message being edited
+    const [isItemBeingEdited, setIsItemBeingEdited] = useState<boolean>(false)
```
[the remark this answers is not in the document]
Comment (abeatrix) [reply lib/ui/src/chat/TranscriptItem.tsx:108]: addessed in https://github.com/sourcegraph/cody/pull/2808!

File: lib/ui/src/chat/TranscriptItem.tsx (hunk 2 above)
Comment (abeatrix) [reply lib/ui/src/chat/TranscriptItem.tsx:115]: Covered by the e2e test to ensure all edit items are hidden when an edit is happening :)

**Comment by abeatrix** [comment]
more works are being done in a newer branch for the latest change in the figma design 🚧 

https://github.com/sourcegraph/cody/pull/2811

File: lib/ui/src/chat/TranscriptItem.tsx [hunk 8]
```diff
@@ -95,46 +97,62 @@ export const TranscriptItem: React.FunctionComponent<
...
-                    value={formInput}
-                    autoFocus={true}
-                    required={true}
-                    onInput={event => setFormInput((event.target as HTMLInputElement).value)}
-                    onKeyDown={event => {
-                        if (event.key === 'Escape') {
-                            setBeingEdited(false)
-                        }
+    // Only returns command name if it is the first word in the message to remove markdown links
+    const initValue = message.displayText?.startsWith('/')
```
[the remark this answers is not in the document]
Comment (abeatrix) [reply lib/ui/src/chat/TranscriptItem.tsx:101]: It is currently not supported, but we can add the tests once https://github.com/sourcegraph/cody/pull/2473 is merged 👍

File: agent/src/index.test.ts [hunk 9]
```diff
@@ -647,10 +647,14 @@ describe('Agent', () => {
             }
         })
 
+        if (process.env.CODY_NODE_VERSION === '16') {
+            it('Skipping edit chat test in Node.js v16 because of unsupported API used', () => {})
+            return
+        }
+
         it(
```
Comment (olafurpg) [inline agent/src/index.test.ts:655]: FYI You can import this helper here and use it like this https://sourcegraph.com/github.com/sourcegraph/cody@8968a9b51794e046781c7c98bc8aff337e4fb429/-/blob/agent/src/cli/evaluate-autocomplete/AutocompleteMatcher.test.ts?L18%3A10-18%3A18
```suggestion
        it.skipIf(isNode16())(
```

**Comment by abeatrix** [comment]
repalced by https://github.com/sourcegraph/cody/pull/2826

**Pull request closed without merging**